plume-common/src/activity_pub/inbox.rs

```diff
@@ -1,48 +1,439 @@
+        A: AsActor<&'a C, Error=E>,
+        V: activitypub::Activity,
+        M: AsObject<A, V, O, &'a C, Error=E>,
+        M::Output: Into<R>,
+        O: activitypub::Object,
+    {
+        match self {
+            Inbox::NotHandled(ctx, act, e) => {
+                if serde_json::from_value::<V>(act.clone()).is_ok() {
+                    let actor = match A::get_or_fetch(ctx, get_id(act["actor"].clone())) {
+                        Ok(a) => a,
+                        // If the actor was not found, go to the next handler
+                        Err(_) => return Inbox::NotHandled(ctx, act, InboxError::InvalidActor),
+                    };
+
+                    let obj: O = match serde_json::from_value(act["object"].clone()) {
+                        Ok(o) => o,
+                        // If the object was not of the expected type, try to dereference it
+                        // and if it is still not valid, go to the next handler
+                        Err(_) => match reqwest::get(&get_id(act["object"].clone()))
+                                .map_err(|_| InboxError::DerefError)
+                                .and_then(|mut r| r.json().map_err(|_| InboxError::InvalidObject))
+                        {
+                            Ok(o) => o,
+                            Err(err) => return Inbox::NotHandled(ctx, act, err),
+                        }
+                    };
+
+                    match M::activity(ctx, actor, obj, &get_id(act.clone())) {
+                        Ok(res) => Inbox::Handled(res.into()),
+                        Err(e) => Inbox::Failed(e)
+                    }
+                } else {
+                    // If the Activity type is not matching the expected one for
+                    // this handler, try with the next one.
+                    Inbox::NotHandled(ctx, act, e)
+                }
+            },
+            other => other
+        }
+    }
+
+    /// Transforms the inbox in a `Result`
+    pub fn done(self) -> Result<R, E> {
+        match self {
+            Inbox::Handled(res) => Ok(res),
+            Inbox::NotHandled(_, _, err) => Err(E::from(err)),
+            Inbox::Failed(err) => Err(err),
+        }
+    }
+}
 
-    fn try_from_activity(conn: &C, act: Create) -> Result<Self, Self::Error> {
-        Self::from_activity(
-            conn,
-            act.create_props.object_object()?,
-            act.create_props.actor_link::<Id>()?,
-        )
+/// Get the ActivityPub ID of a JSON value.
+///
+/// If the value is a string, its value is returned.
+/// Otherwise we return its `id` field, if it is a string
+///
+/// # Panics
+///
+/// This function panics if the value is neither a string nor an object with an
+/// `id` field that is a string.
+fn get_id<'a>(json: serde_json::Value) -> String {
+    match json {
+        serde_json::Value::String(s) => s,
+        serde_json::Value::Object(map) => get_id(map["id"].clone()),
```

Comment:
is it calling itself recursively? a `{"id":{"id":{"id":{"id":"something"}}}` would be considered valid and having an id if so, and as it is cloning, with a big enough json, memory usage could grow quadratic to the size of the original payload